lib/src/facts/aix/processor_resolver.cc

```diff
@@ -0,0 +1,143 @@
+
+        vector<string> processor_types;
+        PdDv* predefined_processor = static_cast<PdDv*>(odm_get_first(PdDv_class, const_cast<char*>("class=processor"), nullptr));
+        while (predefined_processor) {
+            check_odm_return(predefined_processor, "could not fetch processor types from PvDv");
+            LOG_DEBUG("got a processor type: %1%", predefined_processor->uniquetype);
+            processor_types.push_back(predefined_processor->uniquetype);
+            free(predefined_processor);
+            predefined_processor = static_cast<PdDv*>(odm_get_next(PdDv_class, nullptr));
+        }
+
+        vector<string> processor_names;
+        for (string& type : processor_types) {
+            string query = (boost::format("PdDvLn=%1%") % type).str();
+            CuDv* processor = static_cast<CuDv*>(odm_get_first(CuDv_class, const_cast<char*>(query.c_str()), nullptr));
+            while (processor) {
+                check_odm_return(processor, "could not fetch processors from CuDv");
+                LOG_DEBUG("got a processor: %1%", processor->name);
+                processor_names.push_back(processor->name);
+                free(processor);
+                processor = static_cast<CuDv*>(odm_get_next(CuDv_class, nullptr));
+            }
+        }
+
+        for (string& name : processor_names) {
+            string query = (boost::format("name=%1%") % name).str();
+            physical_processor proc;
+            CuAt* attribute = static_cast<CuAt*>(odm_get_first(CuAt_class, const_cast<char*>(query.c_str()), nullptr));
+            while (attribute) {
+                check_odm_return(attribute, "Could not fetch processor attributes from CuAt");
+                LOG_DEBUG("got attribute %1%=%2% for processor %3%", attribute->attribute, attribute->value, name);
+                if (attribute->attribute == string("frequency")) {
+                    proc.frequency = stoll(attribute->value);
+                } else if (attribute->attribute == string("type")) {
+                    proc.type = attribute->value;
+                } else if (attribute->attribute == string("smt_threads")) {
+                    proc.smt_threads = stoi(attribute->value);
+                } else if (attribute->attribute == string("smt_enabled")) {
+                    proc.smt_enabled = (attribute->value == string("true"));
+                } else {
+                    LOG_INFO("don't know what to do with processor attribute %1%", attribute->attribute)
+                }
+                free(attribute);
+                attribute = static_cast<CuAt*>(odm_get_next(CuAt_class, nullptr));
+            }
+
+            result.physical_count++;
+
+            if (result.speed == 0) {
+                result.speed = proc.frequency;
+            } else if (result.speed != proc.frequency) {
+                LOG_WARNING("mismatched processor frequencies found; facter will only report one of them");
```

Comment:
I have no idea if AIX supports heterogenous processors or not, but it's an easy enough check/warning to leave in just in case. If a user reports it we can figure out how to represent it properly